lib/hw_unique_key/hw_unique_key.c

```diff
@@ -5,57 +5,234 @@
 	err = fprotect_area_no_access(huk_addr, CONFIG_FPROTECT_BLOCK_SIZE);
 	if (err != 0) {
 		printk("Fprotect failed with error code: %d\n\r", err);
-		NVIC_SystemReset();
+		k_panic();
 	}
 
 	printk("The Hardware Unique Key loaded successfully!\n\r");
-	return 0;
+}
+
+#else
+
+#define KMU_KEYSLOT_SIZE_WORDS 4
+
+/* Check whether a Hardware Unique Key has been written to the KMU. */
+static bool key_written(uint32_t kmu_slot)
+{
+	uint32_t idx = UICR_KEY_INDEX(kmu_slot);
+
+	NRF_KMU->SELECTKEYSLOT = kmu_slot;
+	if (NRF_UICR_S->KEYSLOT.CONFIG[idx].PERM != 0xFFFFFFFF) {
+		NRF_KMU->SELECTKEYSLOT = 0;
+		return true;
+	}
+	if (NRF_UICR_S->KEYSLOT.CONFIG[idx].DEST != 0xFFFFFFFF) {
+		NRF_KMU->SELECTKEYSLOT = 0;
+		return true;
+	}
+	for (int i = 0; i < KMU_KEYSLOT_SIZE_WORDS; i++) {
+		if (NRF_UICR_S->KEYSLOT.KEY[idx].VALUE[i] != 0xFFFFFFFF) {
+			NRF_KMU->SELECTKEYSLOT = 0;
+			return true;
+		}
+	}
+	NRF_KMU->SELECTKEYSLOT = 0;
+	return false;
+}
+
+
+#if HUK_HAS_CC310
+void hw_unique_key_write(uint32_t kmu_slot, const uint8_t *key)
+{
+	int err = -1;
+
+	if (kmu_slot == HUK_KEYSLOT_KDR) {
+		err = nrf_cc3xx_platform_kmu_write_kdr_slot(key);
+	} else {
+		err = nrf_cc3xx_platform_kmu_write_key_slot(
+			UICR_KEY_INDEX(kmu_slot), NRF_CC3XX_PLATFORM_KMU_AES_ADDR,
+			0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk, key);
+	}
+
+	if (err != 0) {
+		printk("The HUK writing (%d) failed with error code: %d\n\r",
+			err, kmu_slot);
+		k_panic();
+	}
+}
+
+bool hw_unique_key_written(uint32_t kmu_slot)
+{
+	return key_written(kmu_slot);
+}
+
+void hw_unique_key_load_kdr(void)
+{
+	int err = nrf_cc3xx_platform_kmu_push_kdr_slot_and_lock();
+
+	if (err != 0) {
+		printk("The HUK loading failed with error code: %d\n\r", err);
+		k_panic();
+	}
+}
+
+
+#elif HUK_HAS_CC312
+void hw_unique_key_write(uint32_t kmu_slot, const uint8_t *key)
+{
+	int err = nrf_cc3xx_platform_kmu_write_key_slot(
+			UICR_KEY_INDEX(kmu_slot), NRF_CC3XX_PLATFORM_KMU_AES_ADDR,
+			0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk, key);
+	if (err == 0) {
+		err = nrf_cc3xx_platform_kmu_write_key_slot(
+				UICR_KEY_INDEX(kmu_slot) + 1, NRF_CC3XX_PLATFORM_KMU_AES_ADDR_2,
+				0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk,
+				key + (HUK_SIZE_BYTES / 2));
+	}
+
+	if (err != 0) {
+		printk("The HUK writing (%d) failed with error code: %d\n\r",
+			err, kmu_slot);
+		k_panic();
+	}
+}
+
+bool hw_unique_key_written(uint32_t kmu_slot)
+{
+	return key_written(kmu_slot) || key_written(kmu_slot + 1);
+}
+
+#endif
+#endif
+
+void hw_unique_key_write_random(void)
+{
+	uint8_t key[MBEDTLS_ENTROPY_MAX_GATHER];
+	size_t olen;
+
+	BUILD_ASSERT(MBEDTLS_ENTROPY_MAX_GATHER >= (HUK_SIZE_BYTES * 3));
+
+	for (int i = 0; i < sizeof(slots) / 4; i++) {
+		if (hw_unique_key_written(slots[i])) {
+			printk("One or more keys already set. Cannot overwrite\n\r");
+			k_panic();
+		}
+	}
+
+	mbedtls_platform_context platform_ctx = {0};
+
+	int err = mbedtls_platform_setup(&platform_ctx);
```

Comment:
Dependency on mbed TLS configurations and project is unneccesary.

Use nrf_cc3xx_platform_init (which can be called multiple times without extra effort)